### PR TITLE
Use psutil to retrieve the user experiments

### LINKED ISSUE #2104: autosubmit stop not working with 4.1.12
Just confirmed in the ClimateDT VM. I was running an experiment (`t02z` FWIW), when @ainagaya pointed there could be something wrong with my config. I tried to stop my experiment:

```bash
$ autosubmit stop -f t02z
...
Autosubmit admin user: eadmin is not set
Autosubmit is running with 4.1.12
Expid t02z was not running
```

It says the experiment is not running, which is wrong,

```bash
$ ps aux | grep t02z
brunoki+ 4015584  2.7  0.3 785944 121064 pts/15  Sl   11:00   0:12 ...4.1.12-dev-e326af7/bin/python3.9 ...4.1.12-dev-e326af7/bin/autosubmit run -v t02z
brunoki+ 4015604  0.0  0.2 515080 88840 pts/15   S    11:00   0:00 autosubmit log t02z recovery local
brunoki+ 4015607  0.0  0.2 650076 93876 pts/15   Sl   11:00   0:00 autosubmit log t02z recovery lumi-login
brunoki+ 4015614  0.0  0.2 715612 92192 pts/15   Sl   11:00   0:00 autosubmit log t02z recovery lumi
brunoki+ 4017770  0.0  0.0  12220  1192 pts/15   S+   11:08   0:00 grep --color=auto t02z
```

I also checked the experiment logs, and the main loop was still running and updating the log files. The model was also timestepping and the `SIM` task was running fine (we killed it unnecessarily, but good we found this bug).

Tried without `force`.

```bash
$ autosubmit stop t02z
...
Autosubmit admin user: eadmin is not set
Autosubmit is running with 4.1.12
Expid t02z was not running
Checking the status of the expid: t02z
Expid t02z is stopped
```

Running `--cancel` worked though, at least for `SIM` (`DN` didn't?):

```bash
$ autosubmit stop --cancel t02z
...
...
Autosubmit admin user: eadmin is not set
Autosubmit is running with 4.1.12
Expid t02z was not running
Checking the status of the expid: t02z
Expid t02z is stopped

Checking configuration files...
...
...
[WARNING] Some defined platforms have the following issues: [['LUMI-CLUSTER', 'Mandatory SCRATCH_DIR parameter not found'], ['LOCAL-LOGIN', 'Mandatory SCRATCH_DIR parameter not found']]
Platforms sections: OK
Jobs sections OK
Autosubmit general sections OK
Configuration files OK



Loading JobList
Load finished
Creating jobs...
[WARNING] After calculations, the total sizes are: SplitSize*SplitUnitSize:24 hours, ChunkSize*ChunkUnitsize:744 hours.
For the section DN with date:19900101 the number of splits is 31.
[WARNING] After calculations, the total sizes are: SplitSize*SplitUnitSize:24 hours, ChunkSize*ChunkUnitsize:744 hours.
For the section APP_ENERGY_ONSHORE with date:19900101 the number of splits is 31.
[WARNING] After calculations, the total sizes are: SplitSize*SplitUnitSize:24 hours, ChunkSize*ChunkUnitsize:744 hours.
For the section OPA_ENERGY_ONSHORE with date:19900101 the number of splits is 31.
Adding dependencies to the graph..
Adding dependencies to the job..
...
...
Checking the connection to all platforms in use
[LUMI] Correct user privileges for host lumi.csc.fi
[LUMI] Connection successful to host lumi.csc.fi
[LUMI-LOGIN] Correct user privileges for host lumi.csc.fi
[LUMI-LOGIN] Connection successful to host lumi.csc.fi
Cancelling job t02z_19900101_fc0_1_1_DN on platform LUMI-LOGIN
Changing status of job t02z_19900101_fc0_1_1_DN to FAILED
Cancelling job t02z_19900101_fc0_1_SIM on platform LUMI
Changing status of job t02z_19900101_fc0_1_SIM to FAILED
```

In the end I had to `kill -9` the `autosubmit run t02z` process and its children later (two `kill -9` commands, to avoid orphaned processes as killing `autosubmit run` moved then under the init process).

If it were an operator running it, they would have to be careful to kill the processes, and stop remote jobs.

### PR DESCRIPTION
**Pull request by kinow**
Closes #2104 

- [x] Write unit tests
- [x] Local tests
- [ ] Maybe test in the VM too, or hub, or rocky, etc...

The Autosubmit VM seems to be under heavy load, so I will refrain from doing the tests there for now.

### CODE REVIEW

File: test/unit/test_stop.py [hunk 1]
```diff
@@ -1,118 +1,66 @@
...
+def test_stop_expids(autosubmit, expids: str):
+    """Test that we can successfully stop by expids"""
+    autosubmit.stop(
+        expids,
+        force=True,
+        all_expids=False,
+        force_all=True,
+        cancel=False,
+        current_status="RUNNING",
+        status="FAILED")
```
Comment (kinow) [inline test/unit/test_stop.py:66]: Pending these tests...

File: autosubmit/job/job_utils.py [hunk 2]
```diff
@@ -459,4 +461,57 @@ def get_collection_of_fixes_applied(self):
...
+        Log.info(f'Cancelling job {job.name} on platform {job.platform.name}')
+        try:
+            job.platform.send_command(f'{job.platform.cancel_cmd} {str(job.id)}', ignore_log=True)
+        except Exception as e:
+            Log.warning(f"Failed to cancel job {job.name} on platform {job.platform.name}: {str(e)}")
+
+        Log.info(f"Changing status of job {job.name} to {target_status}")
+        job.status = Status.KEY_TO_VALUE[target_status]
+
+    job_list.save()
```
Comment (kinow) [inline autosubmit/job/job_utils.py:517]: Pending tests for this...

File: test/unit/test_stop.py [hunk 3]
```diff
@@ -1,118 +1,66 @@
...
-import tempfile
-from mock import Mock, patch
-from random import randrange
-from pathlib import Path
-from autosubmit.job.job import Job
-from autosubmit.job.job_common import Status
-from autosubmit.job.job_list import JobList
-from autosubmit.job.job_list_persistence import JobListPersistencePkl
-from autosubmitconfigparser.config.yamlparser import YAMLParserFactory
+"""This file contains the test for the `autosubmit stop` command."""
```
Comment (kinow) [inline test/unit/test_stop.py:19]: s/the test/tests/

**Comment by kinow** [comment]
For next week when I continue working on this: rebase to use latest `psutil`, then test with that version...

File: autosubmit/autosubmit.py [hunk 4]
```diff
@@ -20,7 +20,6 @@
 import requests
 # You should have received a copy of the GNU General Public License
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
-import threading
```
Comment (kinow) [inline autosubmit/autosubmit.py:23]: Not used.

**Comment by kinow** [comment]
Wrote tests for all new functions in other files, and code moved. Also finished the tests for the failure scenarios of the `stop` function. Next is to test that `stop` works.

**Comment by kinow** [comment]
Now just pending to cover some corner cases that Codecov found :pray: 

image

**Comment by kinow** [comment]
Unit tests done now. The missing line reported is showing a line that is not touched, but that received a new line at the end as I added code below it.

image

Now some manual tests, and see if I can test this in the VM...

**Comment by kinow** [comment]
Found a bug in my code running it locally as I wrote it to find `autosubmit run a000` (in this order, but allows for other parameters like `-lc DEBUG`, etc., in between tokens), but I realized it will appear in the process list as `/something/something/bin/autosubmit`, so I'm just getting the name of the `pathlib.Path` object to compare with `autosubmit`.

File: autosubmit/helpers/processes.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    expids: List[str] = []
+    for process in [p for p in process_iter(['pid', 'cmdline', 'username']) if p.username() == user]:
+        expid = _match_autosubmit_cmdline(process.cmdline())
+        if expid:
+            expids.append(expid)
+
+    return expids
+
+
+# TODO: check with Dani if the platform is needed (no processes using it on hub or destine vm?)
```
Comment (kinow) [inline autosubmit/helpers/processes.py:154]: To check with Dani :point_up:

File: autosubmit/helpers/processes.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+from psutil import process_iter
+
+from log.log import Log
+
+"""Functions to handle linux processes."""
+
+# TODO: Small note: in reality expids could have five or more letters.
+#       Even if unlikely...
+_EXPID_REGEX = re.compile(r'''
+    ^[a-zA-Z0-9]{4}$            # An exact 4-characters alpha-numeric string.
```
Comment (kinow) [inline autosubmit/helpers/processes.py:29]: Will update later, "An alpha-numeric string with 4 characters"

File: autosubmit/autosubmit.py [hunk 7]
```diff
@@ -6009,110 +6009,57 @@ def stop(expids, force=False, all=False, force_all=False, cancel=False, current_
...
-                    if not proccess_id(expid):
+                while True:
+                    if not process_id(expid):
                         Log.info(f"Expid {expid} is stopped")
-                        process_end = True
-                    else:
-                        Log.info(f"Waiting for the autosubmit run to safety stop: {expid}")
-                        sleep(5)
+                        break
+                    Log.info(f"Waiting for the autosubmit run to safety stop: {expid}")
```
Comment (kinow) [inline autosubmit/autosubmit.py:6061]: For later: clarify the "safety stop" (from mattermost)

Comment (dbeltrankyl) [reply autosubmit/autosubmit.py:6061]: This means that after prompting control +c, AS will stop the execution either at the end of the loop or at the start of the next one ( after all submission, squeue, etc, is done ) to be sure that the job_list is saved.

On a side note, I checked the job_list saves the other day, and there are many saves precisely because people spam control + c, and the job_list was sometimes not saved in the past.

Comment (dbeltrankyl) [reply autosubmit/autosubmit.py:6061]: more test.. 
 #second run after stopping the last one
 
image

Safety timer set to 0


# thrid run after the second one
image

File: autosubmit/helpers/processes.py (hunk 5 above)
Comment (dbeltrankyl) [reply autosubmit/helpers/processes.py:154]: Hello Bruno, 

I'm not sure what you mean by this. 

>  TODO: check with Dani if the platform is needed (no processes using it on hub or destine VM?)

I mean, the platform is to cancel the job, and I see it there in the cancel_cmd below not here

File: autosubmit/job/job_utils.py [hunk 8]
```diff
@@ -459,4 +461,57 @@ def get_collection_of_fixes_applied(self):
...
+
+    :param job_list: Autosubmit job list object.
+    :type job_list: JobList
+    :param active_jobs_filter: Filter used to identify jobs considered active.
+    :type active_jobs_filter: List[str]
+    :param target_status: Final status of the jobs cancelled.
+    :type target_status: str|None
+    """
+    if not target_status or target_status not in Status.VALUE_TO_KEY.values():
+        raise ValueError(f'Cancellation target status of jobs is not valid: {target_status}')
```
Comment (dbeltrankyl) [inline autosubmit/job/job_utils.py:493]: I'm not sure if this will this be shown as an "unknown error" when triggered due to the code in the `$git_root/autosubmit/__init__.py`:

```python

    if is_autosubmit_error:
        e: Union[AutosubmitError, AutosubmitCritical] = e  # type: ignore
        if e.trace:
            Log.debug("Trace: {0}", str(e.trace))
        Log.critical("{1} [eCode={0}]", e.code, e.message)
        err_code = e.code

    if not is_portalocker_error and not is_autosubmit_error:
        msg = "Unexpected error: {0}.\n Please report it to Autosubmit Developers through Git"
        args = [str(e)]
        Log.critical(msg.format(*args))
        err_code = 7000
 ```

File: autosubmit/helpers/processes.py (hunk 6 above)
Comment (dbeltrankyl) [reply autosubmit/helpers/processes.py:29]: It wasn't updated, right?

File: autosubmit/autosubmit.py [hunk 9]
```diff
@@ -6011,95 +6012,65 @@ def stop(expids, force=False, all=False, force_all=False, cancel=False, current_
...
-            raise AutosubmitCritical("Invalid status. Expected one of {0}".format(Status.VALUE_TO_KEY.keys()), 7011)
-        if "," in current_status:
-            current_status = current_status.upper().split(",")
-        else:
-            current_status = current_status.upper().split(" ")
+            raise AutosubmitCritical(f"Invalid status. Expected one of {Status.VALUE_TO_KEY.keys()}", 7011)
+
         try:
+            current_status = current_status.replace(',', ' ').split(' ')
+            current_status = [status.upper() for status in filter(lambda x: x, current_status)]
```
[the remark this answers is not in the document]
Comment (dbeltrankyl) [reply autosubmit/autosubmit.py:6021]: Did you add the test?

File: autosubmit/job/job_utils.py (hunk 8 above)
Comment (dbeltrankyl) [reply autosubmit/job/job_utils.py:493]: Also..

The experiment must be stopped before canceling the jobs due to consistency ( and I think this is the actual behavior and older one).

But I think we should move this check to the start of the stop function for all expids. 

Because if we're checking here if the cancel is well defined. Autosubmit would stop all experiments and then will hard crash when trying to cancel the jobs.

**Review by dbeltrankyl — APPROVED** [review]
Thanks @kinow 

Awesome work, I find this very complete and detailed.

Left some comments, but we can merge it as it is

**Comment by kinow** [comment]
Rebasing...

**Comment by kinow** [comment]
Reviewed a PR from Dani, did a pass over the PRs open, now fresh :coffee: (mate actually), ready to restart work on this one. Will start by the tests...

File: autosubmit/helpers/processes.py (hunk 6 above)
Comment (kinow) [reply autosubmit/helpers/processes.py]: No. Just did that. Thanks!

File: autosubmit/helpers/processes.py (hunk 5 above)
Comment (kinow) [reply autosubmit/helpers/processes.py:154]: Hi @dbeltrankyl 

I also didn't remember why I wrote this TODO :grimacing: But some searching here, I found the old code.

I should have included it in my comment :facepalm: But it was about this part in the `grep` that searches for a platform. From what I could tell, that's always false, and never used. But I don't know if there were plans to add it later?

image

File: autosubmit/job/job_utils.py (hunk 8 above)
Comment (kinow) [reply autosubmit/job/job_utils.py:498]: >I'm not sure if this will this be shown as an "unknown error" when triggered due to the code in the $git_root/autosubmit/__init__.py:

I followed the code trace back to that function, and I think you are correct, it would not print that and just display an unknown error. I changed the code to raise an `AutosubmitCritical` (I think an invalid target status can cause everything to crash?).

**Comment by kinow** [comment]
Addressed feedback, rebased. I think it's ready for a new review, @dbeltrankyl :+1:

**Comment by kinow** [comment]
Alright, now it's ready :relieved:

**Review by dbeltrankyl — APPROVED** [review]

**Comment by dbeltrankyl** [comment]
Thanks @kinow , for me it can be merged and included in 4.1.12

**Comment by kinow** [comment]
Thanks, @dbeltrankyl , then merging this now :tada:
